include/xgboost/data.h

```diff
@@ -393,35 +403,44 @@ class RowSet {
  *  - Provide a DataSource, that can be passed to DMatrix::Create
  *      This can be used to re-use inmemory data structure into DMatrix.
  */
-class DMatrix {
+class DMatrix final {
  public:
   /*! \brief default constructor */
-  DMatrix()  = default;
+  DMatrix() = delete;
+  /*! \brief copy constructor */
+  DMatrix(const DMatrix&) = delete;
+  /*! \brief assignment operator */
+  DMatrix& operator=(const DMatrix&) = delete;
   /*! \brief meta information of the dataset */
-  virtual MetaInfo& Info() = 0;
+  MetaInfo& Info();
   /*! \brief meta information of the dataset */
-  virtual const MetaInfo& Info() const = 0;
+  const MetaInfo& Info() const;
+  /**
+   * \brief Gets batches. Use range based for loop over BatchSet to access individual batches.
+   */
+  template<typename T>
+  BatchSet<T> GetBatches(const std::string& format);
   /**
    * \brief Gets row batches. Use range based for loop over BatchSet to access individual batches.
    */
-  virtual BatchSet GetRowBatches() = 0;
-  virtual BatchSet GetSortedColumnBatches() = 0;
-  virtual BatchSet GetColumnBatches() = 0;
+  BatchSet<SparsePage> GetRowBatches();
```

Comment:
I imagine eventually these methods would disappear and instead we would have a derivative class of SparsePage unique for each of these scenarios and we would use GetBatches?

Reply:
Yes, the user interface is through `GetBatches<T>()`, ended up still needing these for dispatching, but they are now protected/private, so more or less just implementation detail.